diff --git a/fesvr/device.h b/fesvr/device.h
index 1387b74..c2d7829 100644
--- a/fesvr/device.h
+++ b/fesvr/device.h
@@ -5,6 +5,7 @@
 #include <queue>
 #include <cstring>
 #include <string>
+#include <cstdint>
 #include <functional>
 
 class memif_t;
diff --git a/riscv/decode.h b/riscv/decode.h
index 5b6a0ee..5422b4b 100644
--- a/riscv/decode.h
+++ b/riscv/decode.h
@@ -78,7 +78,7 @@ class insn_t
 public:
   insn_t() = default;
   insn_t(insn_bits_t bits) : b(bits) {}
-  insn_bits_t bits() { return b & ~((UINT64_MAX) << (length() * 8)); }
+  insn_bits_t bits() { int l = length(); return l >= 8 ? b : (b & ~((UINT64_MAX) << (l * 8))); }
   int length() { return insn_length(b); }
   int64_t i_imm() { return int64_t(b) >> 20; }
   int64_t shamt() { return x(20, 6); }
